RX: Add meson.build to enable Renesas RX build

duynguyenxa opened the pull request:
Add meson.build to enable picolibc for rx-zephyr-elf

duynguyenxa committed:
RX: Add meson.build to enable Renesas RX build

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

keith-packard commented:
Is there any emulator we could use to test this code using the picolibc CI infrastructure?

duynguyenxa commented:
hi @keith-packard , there is a support for RX core on qemu, but it seem that it's not well maintained, I'm able to use my local build of the compiler to build some sample application on Zephyr and execute on the qemu_rx environment. Now I'm checking on the Zephyr Kernel test of the RX core.
Is it appropriate to use the qemu_rx for picolibC CI ?

keith-packard commented:
>  Is it appropriate to use the qemu_rx for picolibC CI ?

Yup, that'd be just fine. We have a separate repo in the picolibc project that provides CI tools -- https://github.com/picolibc/picolibc-ci-tools that we would use to build a custom qemu for this target.

KhiemNguyenT commented:
@kartben may you help to assign the suitable person as assignee ?

duynguyenxa committed:
picocrt/rx: Create a real interrupt table

Add RX vector interrupt table at fvector section (0xFFFFFF80)
_start function must be placed at 0xFFFFFFFC, fvector section
must be defined in ld file

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

duynguyenxa committed:
picocrt/machine/rx/: Add support for crt0.c for RX target

Add stack intialization instruction in crt0.c and jump
to _cstart

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

duynguyenxa committed:
semihost/rx Supporting semihosting for RX qemu

Adding semihosting method by using SCI0 for RX qemu

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

duynguyenxa committed:
scripts: Support script to execute RX build and test

Add configure file, build script and run script for RX zephyr
elf compiler

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

duynguyenxa commented:
Hi @keith-packard , I've just updating commits to add basic crt and semihosting method for this compiler, I met some issue with the test run as it's seem that semihosting is not working yet so all the test execution was timeout, I'm still checking on that. It may take some time as I'm not very experience with these stuff. I would very appreciate your help on this. Thank you

keith-packard commented:
> Hi @keith-packard , I've just updating commits to add basic crt and semihosting method for this compiler, I met some issue with the test run as it's seem that semihosting is not working yet so all the test execution was timeout, I'm still checking on that. It may take some time as I'm not very experience with these stuff. I would very appreciate your help on this. Thank you

Yeah, looking at the qemu code it seems there isn't any semihosting support for RX. That's OK; picolibc's "semihosting" layer is quite flexible and has several examples that work with only a serial port, which the qemu RX port appears to include.

Take a look at the or1k, sh or sparc semihost support in picolibc. Those all access an emulate serial port to send and receive data through the qemu instance. The sparc code also uses that to terminate the qemu process by using the `monitor-e9` script which checks for a 0xe9 character in the output stream and extracts an exit code after it sees that.

 * https://github.com/picolibc/picolibc/tree/main/semihost/machine/or1k
 * https://github.com/picolibc/picolibc/tree/main/semihost/machine/sh
 * https://github.com/picolibc/picolibc/tree/main/semihost/machine/sparc
 * https://github.com/picolibc/picolibc/blob/main/scripts/monitor-e9

The Renesas SCI serial port emulation code is provided in qemu, you can use that code to figure out how to write a driver which will probably look a lot like the drivers in the above examples.

 * https://github.com/qemu/qemu/blob/master/hw/char/renesas_sci.c

keith-packard commented:
I read through the renesas SCI emulator in qemu and created an untested semihosting implementation that might provide a useful starting point. Take a look at this: https://github.com/picolibc/picolibc/tree/rx-semihost

duynguyenxa commented:
hi @keith-packard , Thank you for helping me, I'm able to use your code and test with helloworld sample, it worked as expected. But for the test, it's fail to build the newlib/libm/test/math_test due to flash size limitation of qemu system.
![image](https://github.com/user-attachments/assets/4ebe157e-0991-4f4a-945c-7cd52c6e1c97)

duynguyenxa commented:
I tried modified source code of RX qemu target to extend flash and RAM, able to build the test, but cannot get it to auto run, issue is qemu-system-rx does not automatically loading and execute the binary file, it required to run a gdb session to execute "load" and "continue" command.
e.g.  I have to manually do this to execute the test:
$ qemu-system-rx -machine gdbsim-r5f562n8 -cpu rx62n -nographic -bios build-rx-zephyr-elf/test/printf-tests -s -S
Start another teminal:
$ rx-zephyr-elf-gdb ../build-rx-zephyr-elf/test/printf-tests --batch -ex "target remote localhost:1234" -ex "load" -ex "continue"

keith-packard commented:
I've pushed my qemu, gcc and picolibc patches which now pass picolibc testing. I found a bug in GCC when generating the scmpu instruction which caused a lot of tests to fail, and then discovered that QEMU has a bug which means we need to run one-instruction-per-block.

We don't need to expand qemu memory at all; there's 16MB available at 0x01000000.

I'm getting picolibc-ci-tools with RX support going, once that's available, I'll post the picolibc PR.

keith-packard commented:
I've got picolibc-ci-tools working and picolibc tests passing now. The ci-tools changes should be easily ported to sdk-ng. I do have questions about which multi-lib variants the toolchain should support. In particular, do we need PID or big-endian-data variants in multilib?

keith-packard commented:
ok, found (and fixed) the qemu bug; patches have been sent upstream. It was some mis-labeled TCG helper functions. With that fixed, the picolibc test suite passes using the regular qemu execution mode. It's a lot faster this way.

duynguyenxa commented:
> I've got picolibc-ci-tools working and picolibc tests passing now. The ci-tools changes should be easily ported to sdk-ng. I do have questions about which multi-lib variants the toolchain should support. In particular, do we need PID or big-endian-data variants in multilib?

Huge thank for your help, currently I have no plan to support big endian yet, about PID, I'm not sure what it's either, and it seem that Renesas does not properly support it as well the the official product, so they are not needed for now

keith-packard commented:
> Huge thank for your help, currently I have no plan to support big endian yet, about PID, I'm not sure what it's either, and it seem that Renesas does not properly support it as well the the official product, so they are not needed for now

I figured out PID mode -- it uses register-offset based addressing for read-only data. That's all working now. I'm building the big-endian-data variants, but the qemu model doesn't support it so those aren't getting tested. The GCC patch to fix use of string instructions has landed upstream; I'm working on a slightly better version which may land at some point. No ETA on the qemu fixes, but we can pull those from picolibc-ci-tools into the qemu for Zephyr as needed.

So, upstream picolibc will have this landed most likely today. After that, we can pull it into Zephyr SDK and pend that for the upcoming release there. I've already got a pending PR for Picolibc 1.8.9, which was just released, so we should probably plan on just pulling picolibc by SHA.

duynguyenxa commented:
@keith-packard , Thank you very much, I'm very appreciate your help for enabling this architecture on picolibC. I close this as the support is updated on main picolibc
